### PR TITLE
Omp offload features

### PR DESCRIPTION
**Pull request by karlqi2000**
# Adding a New Sample(s)

## Description

These samples show new OpenMP Offload Compiler

## Checklist
Administrative
- [ ] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): Paul Peterson

Code Development
- [X] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [X] Adhere to readme template 
- [X] Enforce format via clang-format config file
- [X] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [X] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [X] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see @JoeOster for assistance)
- [X] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [X] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Joe Oster(@JoeOster)
- [X] Tested using Dev Cloud when applicable

### CODE REVIEW

**Review by JoeOster — DISMISSED** [review]
approved for ci only

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/class_member_functor.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+
+#include <bits/stdc++.h>
+using namespace std;
+
+// A Functor
+class inc {
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/class_member_functor.cpp:11]: Types should be Upper case so here inc -> Inc

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/user_defined_mapper.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
+//==============================================================
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+
+extern "C" int printf(const char *, ...);
+struct C {
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/user_defined_mapper.cpp:8]: Better to use a more descriptive name then "struct C".    This get's confused with the extern "C".

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/user_defined_mapper.cpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  }
+  for (i = 0; i < num; ++i)
+    printf("%s%3d %s", (i == 0 ? "Out: " : ""), c_one.arr[i],
+           (i == num - 1 ? "\n" : ""));
+}
+
+int main() {
+  int arr4[] = {1, 2, 4, 8};
+  int arr8[] = {1, 2, 4, 8, 16, 32, 64, 128};
+  int arr4one[] = {1, 2, 4, 8};
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/user_defined_mapper.cpp:52]: arr4one should be snake_case, maybe line arr4_one, same with arr8one -> arr8_one.   If this is a result, then maybe _res of _result instread of _one.

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+#include <iostream>
+
+extern "C" void *omp_target_get_context(int);
+
+#ifdef OCL_BACKEND
+#pragma omp requires unified_shared_memory
+#endif
+
+int main() {
+  const unsigned Size = 200;
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:19]: Magic constants  would have a k prefix kSize = 200.

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  int D = omp_get_default_device();
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:20]: normal variables should be lower case. D -> d

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+#ifdef OCL_BACKEND
+  cl::sycl::queue Q(
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:22]: Don't use the cl:: prefix in DPC++

The Q should be q

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      cl::sycl::context(static_cast<cl_context>(omp_target_get_context(D))),
+      cl::sycl::gpu_selector());
+#else
+  cl::sycl::queue Q;
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:26]: Q should be q

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+#endif
+  std::cout << "SYCL: Running on "
+            << Q.get_device().get_info<cl::sycl::info::device::name>() << "\n";
+  if (!Q.get_device()
+           .get_info<cl::sycl::info::device::usm_shared_allocations>()) {
+    std::cout << "SYCL: USM is not available\n";
+    return 0;
+  }
+  auto validate = [](int *Data) {
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:35]: Data should be data or other snake_case variable.

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    for (unsigned I = 0; I < Size; ++I)
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:36]: I should be i or other snake_case variable.

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+      if (Data[I] != 100 + I) return "failed";
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:37]: The return from main should be 0 (succes) or non-zero (failure).   You should not be passing character array pointers.

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    return "passed";
+  };
+  auto testOmp = [&](int *Data) {
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:40]: Functions and Types should be MixedCase here testOmp -> TestOmp

Data -> data

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    std::fill_n(Data, Size, -1);
+#pragma omp target parallel for device(D)
+    for (unsigned I = 0; I < Size; ++I) {
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:43]: I should be i

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+      Data[I] = 100 + I;
+    }
+    return validate(Data);
+  };
+  auto testDpc = [&](int *Data) {
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:48]: testDpc -> TestDPCPP

Data -> data

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    std::fill_n(Data, Size, -1);
+    Q.submit([&](cl::sycl::handler &cgh) {
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:50]: Don't use cl:: with DPC++
Q -> q
cgh -> h

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+      cgh.parallel_for<class K>(cl::sycl::range<1>(Size),
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:51]: Don't use <class K> unless you are using the "K" as part of a tutorial for profiling or debugging.

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+                                [=](cl::sycl::id<1> I) { Data[I] = 100 + I; });
+    });
+    Q.wait();
+    return validate(Data);
+  };
+
+  int *ompMem = (int *)omp_target_alloc_shared(Size * sizeof(int), D);
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:58]: ompMem -> omp_mem
dpcMem -> dpcpp_mem]

The example is a bit confusing about DPC++ vs SYCL

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+  int *dpcMem = cl::sycl::malloc_shared<int>(Size, Q);
+  std::cout << "SYCL and OMP memory: " << testDpc(ompMem) << "\n";
+  std::cout << "OMP and OMP memory:  " << testOmp(ompMem) << "\n";
+  std::cout << "OMP and SYCL memory: " << testOmp(dpcMem) << "\n";
+  std::cout << "SYCL and SYCL memory: " << testDpc(dpcMem) << "\n";
+  omp_target_free(ompMem, D);
+  cl::sycl::free(dpcMem, Q);
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:65]: Don't use cl:: for DPC++

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp (hunk 15 above)
Comment (karlqi2000) [reply DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp]: With icpx -fsycl, seems like if I remove "<class K>" , i get the following:
/opt/intel/oneapi/compiler/2021.3.0/linux/bin/../include/sycl/CL/sycl/handler.hpp:994:50: error: unnamed lambda '(lambda at usm_and_composability_with_dpcpp.cpp:54:43)' used
    kernel_parallel_for<KernelName, ElementType>(KernelFunc);

**Review by JoeOster — DISMISSED** [review]
approve for ci only

**Review by tomlenth — APPROVED** [review]

**Comment by JoeOster** [comment]
@pmpeter1 -can you update the comments or approve please

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+  sycl::queue q;
+#endif
+  std::cout << "SYCL: Running on "
+            << q.get_device().get_info<sycl::info::device::name>() << "\n";
+  if (!q.get_device()
+           .get_info<sycl::info::device::usm_shared_allocations>()) {
+    std::cout << "SYCL: USM is not available\n";
+    return 0;
+  }
+  auto Validate = [](int *data) {
```
Comment (pmpeter1) [inline DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:37]: You have int main(), but you are returning a char* value from Validate and then return this from main.

This should be type correct.   You should be printing passed/failued, and then returning 0/1.

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

File: DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp (hunk 18 above)
Comment (karlqi2000) [reply DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:37]: Hi @pmpeter1 ,
In the current code,  TestOmp and TestDPCPP are returning the char* value from Validate which gets printed in main(). main() only returns 0.

Comment (JoeOster) [reply DirectProgramming/C++/CompilerInfrastructure/OpenMP_Offload_Features/src/usm_and_composability_with_dpcpp.cpp:37]: @pmpeter1 are you we good with this reply?

**Review by pmpeter1 — APPROVED** [review]

**Pull request merged**